src/main/scala/tech/cryptonomic/conseil/tezos/TezosNodeOperator.scala

```diff
@@ -103,158 +107,170 @@ class TezosNodeOperator(node: TezosRPCInterface) extends LazyLogging {
+        recursiveMetadataFetch(network, pred, minLevel, maxLevel, followFork, collected)
+      }
+    else if (aboveMin)
+      getMetadata(network, current.header.predecessor) flatMap { pred =>
+        recursiveMetadataFetch(network, pred, minLevel, maxLevel, followFork, current :: collected)
+      }
+    else
+      Future.successful(List.empty)
+
+  }
 
   /**
-    * Traverses Tezos blockchain as parameterized.
-    * @param network    Which Tezos network to go against
-    * @param hash       Current block to fetch
-    * @param minLevel   Minimum level at which to stop
-    * @param maxLevel   Level at which to start collecting blocks
-    * @param blockSoFar Blocks collected so far
-    * @param followFork If the predecessor of the minLevel block appears to be on a fork, also capture the blocks on the fork.
-    * @return           All collected blocks
+    * Traverses Tezos blockchain as parameterized, without waiting for the answer
+    *
+    * @param network       Which Tezos network to go against
+    * @param startingBlock Current block to start from
+    * @param minLevel      Minimum level at which to stop
+    * @param maxLevel      Level at which to start collecting blocks
+    * @param followFork    If the predecessor of the minLevel block appears to be on a fork, also capture the blocks on the fork.
+    * @return              All collected blocks, wrapped in a [[Future]]
     */
-  private def processBlocks(
-                             network: String,
-                             hash: String,
-                             minLevel: Int,
-                             maxLevel: Int,
-                             blockSoFar: List[Block] = List[Block](),
-                             followFork: Boolean
-                           ): List[Block] =
-    getBlock(network, hash) match {
-      case Success(block) =>
-        logger.info(s"Current block height: ${block.metadata.header.level}")
-        if(block.metadata.header.level == 0 && minLevel <= 0)
-          block :: blockSoFar
-        else if(block.metadata.header.level == minLevel && !followFork)
-          block :: blockSoFar
-        else if(block.metadata.header.level <= minLevel && followFork)
-          ApiOperations.fetchBlock(block.metadata.header.predecessor) match {
-            case Success(_)     => block :: blockSoFar
-            case Failure(_)     => processBlocks(network, block.metadata.header.predecessor, minLevel, maxLevel, block :: blockSoFar, followFork)
-          }
-        else if(block.metadata.header.level > maxLevel)
-          processBlocks(network, block.metadata.header.predecessor, minLevel, maxLevel, blockSoFar, followFork)
-        else if(block.metadata.header.level > minLevel)
-          processBlocks(network, block.metadata.header.predecessor, minLevel, maxLevel, block :: blockSoFar, followFork)
-        else
-          List[Block]()
-      case Failure(e) => throw e
-    }
+  private def recursiveBlockFetch(
```

Comment:
what gave you the idea to rewrite these? Were you trying to go for speedups? If you want to try that, I would say look at the Refactor ProcessBlocksPr and try and give advice there. It's a separate, tricky issue

Reply:
*ProcessBlocks PR

Reply:
Indeed, I was doing a different kind of rewrite, though.

As I explain in previous comment, this part of the code is here mostly to get a feedback, as I didn't expect it to be merged in the end.

I will strip this changes before removing the WIP label. The part that I really cared about, was the one to load accounts.